### ISSUE TITLE
Android: Restoring purchases doesn't work

### ISSUE THREAD

**Issue by Fox9MonoGame** [jamesmontemagno/InAppBillingPlugin#194]
Restoring consumable and non-consumable in-app puchases isn't working on Android. I have tested it on my Nokia 7 Plus device. I uploaded an app with a consumable and a non-consumable in-app product to Google Play Console and after that I published the app as alpha version and then I downloaded the app from Google Play to test it on my device. Purchasing the in-app products is working, but when I exit the app and then reopen the app, then the purchased products are not getting restored. The same happens when I reinstall the application.
I have only two in-app products in my application, "consumable1" and "nonconsumable1".

Could you explain me what I'm doing wrong? I use almost the same code as in your documentation.

I run this code after the application was loaded,  but the consumable and non-consumable products are never getting restored.

      ```
       public async Task<bool> WasItemPurchased(string productId)
        {
              var wasitempurchased = false;
              var billing = CrossInAppBilling.Current;
              try
              {
                var connected = await billing.ConnectAsync(ItemType.InAppPurchase);

                if (connected == true)
                {
                    //check purchases
                    var purchases = await billing.GetPurchasesAsync(ItemType.InAppPurchase);

                    //check for null just incase
                    if (purchases?.Any(p => p.ProductId == productId) ?? false)
                    {
                        //Purchase restored
                        wasitempurchased = true;
                        if (productId == "consumable1")
                            Coins += 50;
                        if (productId == "nonconsumable1")
                            LicensePurchased = true;
                    }
                    else
                    {
                        //no purchases found
                        wasitempurchased = false;
                    }
                }
                else
                    wasitempurchased = false;

            }

            catch (InAppBillingPurchaseException purchaseEx)
            {
                //Billing Exception handle this based on the type
                wasitempurchased = false;
                Debug.WriteLine("Error: " + purchaseEx);
            }
            catch (Exception ex)
            {
                //Something has gone wrong
                wasitempurchased = false;
                Debug.WriteLine("Issue connecting: " + ex);
            }
            finally
            {
                await billing.DisconnectAsync();
            }

            //This does all the disposing you need
            CrossInAppBilling.Dispose();

            return wasitempurchased;
        }
```

**Comment by Fox9MonoGame** [comment]
I did another test with the Android emulator(API 27). I added some breakpoints to my code and I found out that this part of the code is never executed:

              ```
                {
                    //Purchase restored
                    wasitempurchased = true;
                    if (productId == "consumable1")
                        Coins += 50;
                    if (productId == "nonconsumable1")
                        LicensePurchased = true;
                }
```

**Comment by stevehurcombe** [comment]
@Fox9MonoGame I'm watching this repository because I aim to use this library in the future, so I can't offer specific advice yet, but I have had experience with resolving these types of issues in other scenarios.

What I've noticed is that Visual Studio (on Windows at least) silently ignores exceptions inside async...await patterns and that's why that await on ```GetPurchasesAsync``` never returns. Your challenge is to find out why.

In the past I've downloaded the source (thank you for everyone putting their code on GitHub!) and added it as a project in my own solution. Then I can either eyeball where the exception might be - we've all tried eye-ball debugging - or place my own breakpoints lower down in their code. 

Ideally these libraries should have deep logging to help diagnose issues (maybe it does??).

Sometimes this works, sometimes it doesn't, but if you need a way forward this might help.

Sorry I can't offer more direct advice.

**Comment by Fox9MonoGame** [comment]
I found the issue. The problem was right after I purchased an in-app product.
I called "var consumedItem = await billing.ConsumePurchaseAsync(purchase.ProductId, purchase.PurchaseToken);" after "var purchase = await billing.PurchaseAsync(productId, ItemType.InAppPurchase, payload, verify);".

Restoring in-app products works fine if you don't call "var consumedItem = await billing.ConsumePurchaseAsync(purchase.ProductId, purchase.PurchaseToken);" after you did a purchase.
